components/director/internal/domain/formation/service.go

```diff
@@ -180,6 +205,160 @@ func (s *service) UnassignFormation(ctx context.Context, tnt, objectID string, o
+	}
+
+	err = s.EnsureScenarioAssigned(ctx, in)
+	if err != nil {
+		return model.AutomaticScenarioAssignment{}, errors.Wrap(err, "while assigning scenario to runtimes matching selector")
+	}
+
+	return in, nil
+}
+
+func (s *service) validateThatScenarioExists(ctx context.Context, in model.AutomaticScenarioAssignment) error {
+	availableScenarios, err := s.getAvailableScenarios(ctx, in.Tenant)
+	if err != nil {
+		return err
+	}
+
+	for _, av := range availableScenarios {
+		if av == in.ScenarioName {
+			return nil
+		}
+	}
+
+	return apperrors.NewNotFoundError(resource.AutomaticScenarioAssigment, in.ScenarioName)
+}
+
+func (s *service) getAvailableScenarios(ctx context.Context, tenantID string) ([]string, error) {
+	if err := s.labelDefService.EnsureScenariosLabelDefinitionExists(ctx, tenantID); err != nil {
+		return nil, errors.Wrap(err, "while ensuring that `scenarios` label definition exist")
+	}
+
+	out, err := s.labelDefService.GetAvailableScenarios(ctx, tenantID)
+	if err != nil {
+		return nil, errors.Wrap(err, "while getting available scenarios")
+	}
+	return out, nil
+}
+
+// DeleteAutomaticScenarioAssignment deletes the assignment for a given scenario in a scope of a tenant
+func (s *service) DeleteAutomaticScenarioAssignment(ctx context.Context, in model.AutomaticScenarioAssignment) error {
+	tenantID, err := tenant.LoadFromContext(ctx)
+	if err != nil {
+		return errors.Wrap(err, "while loading tenant from context")
+	}
+
+	if err = s.repo.DeleteForScenarioName(ctx, tenantID, in.ScenarioName); err != nil {
+		return errors.Wrap(err, "while deleting the Assignment")
+	}
+
+	if err = s.RemoveAssignedScenario(ctx, in); err != nil {
+		return errors.Wrap(err, "while unassigning scenario from runtimes")
+	}
+
+	return nil
+}
+
+func (s *service) EnsureScenarioAssigned(ctx context.Context, in model.AutomaticScenarioAssignment) error {
+	runtimes, err := s.runtimeRepo.ListAll(ctx, in.TargetTenantID, nil)
+
+	if err != nil {
+		return errors.Wrapf(err, "while fetching runtimes in target tenant: %s", in.TargetTenantID)
+	}
+	for _, runtime := range runtimes {
+		_, err = s.AssignFormation(ctx, in.Tenant, runtime.ID, graphql.FormationObjectTypeRuntime, model.Formation{Name: in.ScenarioName})
+		if err != nil && !apperrors.IsNotUniqueError(err) {
+			return errors.Wrapf(err, "while assigning formation")
+		}
+	}
+	return nil
+}
+
+// RemoveAssignedScenario removes all the scenarios that are coming from the provided ASA
+func (s *service) RemoveAssignedScenario(ctx context.Context, in model.AutomaticScenarioAssignment) error {
+	runtimes, err := s.runtimeRepo.ListAll(ctx, in.TargetTenantID, nil)
+	if err != nil {
+		return errors.Wrapf(err, "while fetching runtimes in target tenant: %s", in.TargetTenantID)
+	}
+
+	for _, runtime := range runtimes {
+		_, err = s.UnassignFormation(ctx, in.Tenant, runtime.ID, graphql.FormationObjectTypeRuntime, model.Formation{Name: in.ScenarioName})
+		if err != nil && !apperrors.IsNotFoundError(err) {
```

Comment:
In which scenario a not found error will be returned here?

Reply:
I presume none, while I was debugging before opening a PR I assumed the issue could be caused when trying to delete a scenario that doesn't exists, and tried to make the deletion more robust to find out if this is the cause. I seem to have forgotten to remove it.